misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -98,46 +104,41 @@ internal class SqsJobConsumer @Inject internal constructor(
 
-      val futures = messages.map { SqsJob(queueName, queues, metrics, it) }.map { message ->
-        dispatchThreadPool.submit {
-          metrics.jobsReceived.labels(queueName.value, queueName.value).inc()
-
-          tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
-            // If the incoming job has an original trace id, set that as a tag on the new span.
-            // We don't turn that into the parent of the current span because that would
-            // incorrectly include the execution time of the job in the execution time of the
-            // action that triggered the job        q
-            message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
-              ORIGINAL_TRACE_ID_TAG.set(span, it)
-            }
-
-            // Run the handler and record timing
-            try {
-              val (duration, _) = timed { handler.handleJob(message) }
-              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value, queueName.value)
-            } catch (th: Throwable) {
-              log.error(th) { "error handling job from ${queueName.value}" }
-              metrics.handlerFailures.labels(queueName.value, queueName.value).inc()
-              Tags.ERROR.set(span, true)
-              throw th
+      messages.map { SqsJob(queueName, queues, metrics, it) }.forEach { message ->
+        availableThreadSemaphore.acquire()
+        try {
+          dispatchThreadPool.submit {
+            metrics.jobsReceived.labels(queueName.value, queueName.value).inc()
+
+            tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
+              // If the incoming job has an original trace id, set that as a tag on the new span.
+              // We don't turn that into the parent of the current span because that would
+              // incorrectly include the execution time of the job in the execution time of the
+              // action that triggered the job        q
+              message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
+                ORIGINAL_TRACE_ID_TAG.set(span, it)
+              }
+
+              // Run the handler and record timing
+              try {
+                val (duration, _) = timed { handler.handleJob(message) }
+                metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value)
+              } catch (th: Throwable) {
+                log.error(th) { "error handling job from ${queueName.value}" }
+                metrics.handlerFailures.labels(queueName.value).inc()
+                Tags.ERROR.set(span, true)
+                throw th
+              }
             }
           }
+        } finally {
+          availableThreadSemaphore.release()
         }
       }
-
-      for (future in futures) {
-        try {
-          future.get()
-        } catch (e: ExecutionException) {
-          // the exception was already logged when the dispatched task failed above
-          return Status.FAILED
```

Comment:
also we've lost exponential backoff of the jobs when it's just spinning and failing

Reply:
i mean we have metrics for job failing and returning FAILED. that is no longer happening.

>  i don't think we want to backoff just because a single message in a batch fails processing

typically backing off when things are failing in unexpected ways is a good thing, otherwise, we overload downstream systems.

Reply:
> i mean we have metrics for job failing and returning FAILED. that is no longer happening.

oh, i see. well, we're no longer considering a single failed message to be a failure of the job. as i said, i think this is correct behaviour. we don't want our queue backing off because we're failing to process one message in a batch.

Reply:
> typically backing off when things are failing in unexpected ways is a good thing, otherwise, we overload downstream systems.

in general i'd agree. i think this is an exception, but open to discussing it.

Reply:
one alternative that could work would be to return FAILED iff every task fails. it's a little tricky to implement though with the current architecture (while solving the problem that this PR solves). might be worth it though?

Reply:
thinking about this a little more, i think that the fundamental issue here i think is that our jobqueue processing is single threaded, but we're trying to use it for something that's fundamentally concurrent. 

there's an impedance mismatch here, and the result is that we either:

- lose concurrency
- lose exponential backoff
- reimplement the backoff logic in the queue handling implementation (and probably don't use RepeatedTaskQueue)

thoughts?

Reply:
i'm going to send a PR for option 4, since that's simplest and fairly uncontroversial.